### PR TITLE
[16.0][ADD] delivery_estimated_package_quantity_by_weight

### PR DESCRIPTION
**Pull request by glitchov**
This module computes the amount of packages a picking out should have depending on the weight of the products and the limit fixed by the carrier.
It's fully independent of the delivery_package_number module.

### CODE REVIEW

File: delivery_package_number_theoretical/models/delivery_carrier.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+
+    _inherit = "delivery.carrier"
+
+    maximum_weight_per_package = fields.Float(string="Maximum weight per package")
```
Comment (lmignon) [inline delivery_package_number_theoretical/models/delivery_carrier.py]: @glitchov As it's done for the product form, you should display the uom name of the uom defined for volumes...

```python
    weight_uom_name = fields.Char(string='Weight unit of measure label', compute='_compute_weight_uom_name')

    def _compute_weight_uom_name(self):
        self.weight_uom_name = self.env["product.template"]._get_weight_uom_name_from_ir_config_parameter()
```

File: delivery_package_number_theoretical/views/delivery_carrier_views.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    </record>
+    <record model="ir.ui.view" id="delivery_carrier_tree_view">
+        <field
+            name="name"
+        >delivery.carrier.tree (in delivery_number_package_theoretical)</field>
+        <field name="model">delivery.carrier</field>
+        <field name="inherit_id" ref="delivery.view_delivery_carrier_tree" />
+        <field name="arch" type="xml">
+            <field name="delivery_type" position="after">
+                <field name="maximum_weight_per_package" />
```
Comment (lmignon) [inline delivery_package_number_theoretical/views/delivery_carrier_views.xml]:
```suggestion
                 <label for="weight"/>
                 <div class="o_row" name="maximum_weight_per_package">
                     <field name="maximum_weight_per_package" class="oe_inline"/>
                     <field name="weight_uom_name"/>
                 </div>
```

**Review by lmignon — CHANGES_REQUESTED** [review]
@glitchov Thank you for this addon. Reviewed technically and functionally.  LGTM but we need to display the volume uom configured into odoo next to the `maximum_weight_per_package` field into the delivery.carrier form view to help the user to provide the right value.

**Review by lmignon — APPROVED** [review]
Thank you @glitchov LGTM (Code review + functional tests)

**Review by rousseldenis — APPROVED** [review]
Code review

File: delivery_package_number_theoretical/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            ):
+                rec.is_number_of_packages_visible = True
+            else:
+                rec.is_number_of_packages_visible = False
+
+    @api.depends("is_number_of_packages_visible", "move_ids")
+    def _compute_theoretical_number_of_packages(self):
+        for rec in self:
+            if rec.is_number_of_packages_visible:
+                products_weights = rec.move_ids.mapped("product_id.weight")
```
Comment (jbaudoux) [inline delivery_package_number_theoretical/models/stock_picking.py]: Add in README that product packagings are not supported

File: delivery_package_number_theoretical/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            theoretical_number_of_packages += 1
+            # Try to fit the heaviest product with the lightest.
+            # If it does not work, then the heaviest should have
+            # a box to himself
+            weight = products_weights_list[i] + products_weights_list[j]
+            while weight <= maximum_weight_per_package:
+                i += 1
+                if i < j:
+                    # While the weight of products does not exceed the limit,
+                    # continue adding products in the same package
```
Comment (jbaudoux) [inline delivery_package_number_theoretical/models/stock_picking.py]: Add in README the chosen strategy

File: delivery_package_number_theoretical/models/delivery_carrier.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
```
Comment (jbaudoux) [inline delivery_package_number_theoretical/models/delivery_carrier.py]: Add in README that you don't support multiple delivery packages (stock.package.type) with different max weight for a shipping method

**Review by jbaudoux — COMMENTED** [review]
I find this module quite specific, especially that a specific strategy is applied on a max weight.
I find the module name misleading. Maybe rename to delivery_estimated_package_quantity_by_weight

**Pull request merged**